### DISCUSSION TITLE
Dynamic line indent based on text

### DISCUSSION THREAD

**Discussion by simolus3** [sile-typesetter/sile#1400] [Q&A] [answered]
Apologies if this has already been answered elsewhere, I don't know the proper terminology for this. Essentially, I want to typeset some text with an indent based on some other text, something like the `cc` part of this:

![cc](https://user-images.githubusercontent.com/5738860/170483981-6ac02697-6d55-4071-9ed1-5d64f753e9ee.png)

I've tried some hacks with manual hboxes, but none of them really worked after linebreaks. My current (broken) approach is this:

```lua
SILE.registerCommand("cc", function (_, content)
  SILE.typesetter:leaveHmode()
  SILE.typesetter:typeset("cc: ")
  local offset = SILE.typesetter.frame.state.cursorX
  SILE.settings.temporarily(function()
    SILE.settings.set("document.lskip", SILE.nodefactory.glue(offset))
    SILE.process(content)
  end)
  SILE.typesetter:leaveHmode()
end)
```

I'm then calling it with

```tex
\cc{executive board \cr all members}
```

Which unfortunately doesn't apply the left skip at all:

![bad cc](https://user-images.githubusercontent.com/5738860/170484541-277186c5-36d2-4270-8c0e-4845bc9c20ea.png)

I see how I'm doing something wrong conceptually here, but I'm a bit lost on how to do this properly.

**Comment by alerque** [comment]
The issue with your sample function is that by the time you set the lskip setting the typesetter for the paragraph is already rolling and doesn't look at the setting again.

What you are showing is what is called a hanging indent. What you need to do to make one is in your function in this order:

1. Set the lskip
2. leave hmode so that the lskip gets applied
3. Insert a negative glue the width of the lskip
4. Proceed to typeset your content.
5. At the end, unset the lskip

Of course 1-5 can also all be done inside a temporary settings closure too, the main paint here is that you need to apply that setting, then get out of hmode for it to apply. You also can't set it on the 2nd line, so set it on the whole paragraph and then backtrack for the "hang".

**Comment by alerque** [comment]
Oh ya I should also mention... As of v0.12.1 there is another mechanism for this which might even be better. #1219 and #1264 introduce paragraph shaping that handles hanging indents for N lines. You could use a negative value (e.g. instead of marking out an area for a drop-cap do the inverse) for the indent and N=1 for lines.

**Comment by Omikhleia** [comment, marked as answer]
A possible way (as noted by @alerque other solutions are also doable) using SILE 0.12.5:
```
\begin[papersize=a6]{document}
\script[src=packages/lorem]
\script{
SILE.registerCommand("cc", function (_, content)
  SILE.settings.temporarily(function()
    SILE.call("par")
    -- You probably do not want a paragraph indent, do you?
    SILE.settings.set("document.parindent", SILE.length())
    -- Put the offsetting content in a box for width computation,
    -- but don't output it yet (= steal it from the queue)
    local h = SILE.call("hbox", {}, { "cc: " })
    table.remove(SILE.typesetter.state.nodes) -- steal the box back
    local offset = h.width
    -- Now set the left skip and ensure hmode again
    SILE.settings.set("document.lskip", SILE.nodefactory.glue(offset))
    SILE.typesetter:leaveHmode()
    -- Insert negative offset
    SILE.call("kern", { width = -offset })
    -- Push the box now
    SILE.typesetter:pushHbox(h)
    -- And finally the content
    SILE.process(content)
    SILE.call("par")
  end)
end)
}

\lorem[words=20] 

\cc{executive board \cr all members}

\end{document}
```

![image](https://user-images.githubusercontent.com/18075640/170867092-56b0d55f-401c-43e3-aeb3-bee5941b2889.png)

@simolus3 If this is what you expected, please mark the question as answered. Otherwise, feel free to ask for clarifications.

EDIT N.B. This solution has a drawback if the line is long and gets justified (as the width of the space after "cc:" is variable). Again, there could be several solutions here, depending on taste...

**Reply by simolus3** [reply]
Thanks a lot, this is exactly what I was looking for! Also thank you @alerque for explaining the technical background.
